pkg/provider/vault/auth_iam.go

```diff
@@ -183,3 +143,87 @@ func (c *client) requestTokenWithIamAuth(ctx context.Context, iamAuth *esv1.Vaul
+	if region != "" {
+		return region
+	}
+	return defaultAWSRegion
+}
+
+func (c *client) getAuthMountPathOrDefault(path string) string {
+	if path != "" {
+		return path
+	}
+	return defaultAWSAuthMountPath
+}
+
+func (c *client) getControllerPodCredentials(ctx context.Context, region string, k kclient.Client, jwtProvider util.JwtProviderFactory) (*credentials.Credentials, error) {
+	// First try IRSA (Web Identity Token) - checking if controller pod's service account is IRSA enabled
+	tokenFile := os.Getenv(vaultiamauth.AWSWebIdentityTokenFileEnvVar)
+	if tokenFile != "" {
+		logger.V(1).Info("using IRSA token for authentication")
+		return c.getCredsFromIRSAToken(ctx, tokenFile, region, k, jwtProvider)
+	}
+
+	// Check for Pod Identity environment variables.
+	podIdentityURI := os.Getenv(vaultiamauth.AWSContainerCredentialsFullURIEnvVar)
+	podIdentityToken := os.Getenv(vaultiamauth.AWSContainerAuthorizationTokenFileEnvVar)
+
+	if podIdentityURI != "" && podIdentityToken != "" {
+		logger.V(1).Info("using Pod Identity for authentication")
+		// Return nil to let AWS SDK v1 container credential provider handle Pod Identity automatically
+		return nil, nil
+	}
+
+	// No IRSA or Pod Identity found.
+	return nil, errors.New(errNoAWSAuthMethodFound)
+}
+
+func (c *client) getCredsFromIRSAToken(ctx context.Context, tokenFile, region string, k kclient.Client, jwtProvider util.JwtProviderFactory) (*credentials.Credentials, error) {
+	// IRSA enabled service account, let's check that the jwt token filemount and file exists
+	if _, err := os.Stat(filepath.Clean(tokenFile)); err != nil {
+		return nil, fmt.Errorf(errIrsaTokenFileNotFoundOnPod, tokenFile, err)
+	}
+
+	// everything looks good so far, let's fetch the jwt token from AWS_WEB_IDENTITY_TOKEN_FILE
+	jwtByte, err := os.ReadFile(filepath.Clean(tokenFile))
+	if err != nil {
+		return nil, fmt.Errorf(errIrsaTokenFileNotReadable, tokenFile, err)
+	}
+
+	// let's parse the jwt token
+	parser := jwt.NewParser(jwt.WithoutClaimsValidation())
+
+	token, _, err := parser.ParseUnverified(string(jwtByte), jwt.MapClaims{})
+	if err != nil {
+		return nil, fmt.Errorf(errIrsaTokenNotValidJWT, tokenFile, err) // JWT token parser error
+	}
+
+	var ns string
+	var sa string
+
+	// let's fetch the namespace and serviceaccount from parsed jwt token
+	if claims, ok := token.Claims.(jwt.MapClaims); ok {
+		k8s, ok := claims["kubernetes.io"].(map[string]any)
+		if !ok {
+			return nil, fmt.Errorf(errIrsaTokenNotValidJWT, tokenFile, err)
+		}
+		ns, ok = k8s["namespace"].(string)
+		if !ok {
+			return nil, fmt.Errorf(errIrsaTokenNotValidJWT, tokenFile, err)
+		}
+		saMap, ok := k8s["serviceaccount"].(map[string]any)
+		if !ok {
+			return nil, fmt.Errorf(errIrsaTokenNotValidJWT, tokenFile, err)
+		}
+		sa, ok = saMap["name"].(string)
+		if !ok {
+			return nil, fmt.Errorf(errIrsaTokenNotValidJWT, tokenFile, err)
+		}
+	} else {
+		return nil, fmt.Errorf(errPodInfoNotFoundOnToken, tokenFile, err)
```

Comment:
On top of that this is an incorrect error here, since the actual error is a pod claims assertion error?